File: 073_Set_Matrix_Zeroes.java

// Given a m x n matrix, if an element is 0, set its entire row and column to 0. Do it in place.

// click to show follow up.

// Follow up:
// Did you use extra space?
// A straight forward solution using O(mn) space is probably a bad idea.
// A simple improvement uses O(m + n) space, but still not the best solution.
// Could you devise a constant space solution?

public class Solution {
    public void setZeroes(int[][] matrix) {
        if (matrix == null || matrix[0] == null) return;
        int m = matrix.length, n = matrix[0].length;
        boolean firstRow = false, firstCol = false;
        // traverse the first row
        for (int col = 0; col < n; col ++){
            if (matrix[0][col] == 0){
                firstRow = true;
                break;
            }
        }
        // traverse the first col
        for (int row = 0; row < m; row ++){
            if (matrix[row][0] == 0){
                firstCol = true;
                break;
            }
        }
        // traverse the rest cells in the matrix
        for(int row = 1; row < m; row ++){
            for (int col = 1; col < n; col++){
                if (matrix[row][col] == 0){
                    matrix[0][col] = 0;
                    matrix[row][0] = 0;
                }
            }
        }
        // make row Zero
        for (int row = 1; row < m; row++){
            if (matrix[row][0] == 0){
                makeRowZero(matrix, row);
            }
        }
        // make col Zero
        for(int col = 1; col < n; col++){
            if (matrix[0][col] == 0){
                makeColZero(matrix, col);
            }
        }
        if (firstRow)   makeRowZero(matrix, 0);
        if (firstCol)    makeColZero(matrix, 0);
    }
    
    public void makeRowZero(int[][] matrix, int rowIdx){
        for(int col = 0; col < matrix[0].length; col++){
            matrix[rowIdx][col] = 0;
        }
    }
    public void makeColZero(int[][] matrix, int colIdx){
        for(int row = 0; row < matrix.length; row++){
            matrix[row][colIdx] = 0;
        }
    }
}

File: python/073_Set_Matrix_Zeroes.py
# Given a m x n matrix, if an element is 0, set its entire row
# and column to 0. Do it in place.

# click to show follow up.

# Subscribe to see which companies asked this question

# Show Tags
# Show Similar Problems

# Runtime: 200ms

class Solution(object):
    def setZeroes(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        column_zero = []
        for i, row in enumerate(matrix):
            flag = 0
            for j,num in enumerate(row):
                if num == 0:
                    column_zero.append(j)
                    flag = 1
            if flag:
                matrix[i] = [0 for _ in row]
        for col in column_zero:
            for i,row in enumerate(matrix):
                matrix[i][col] = 0

if __name__ == '__main__':
    sol = Solution()
    matrix = [[1,2,3],
              [2,3,4],
              [3,4,0]]
    sol.setZeroes(matrix)
    print matrix 